### PR TITLE
[MAINT] Use Numpy Random Generator  to replace legacy RandomState

### LINKED ISSUE #4055: [MAINT] Use `np.random.default_rng` for rng fixture
Regarding this, we should actually upgrade the fixture to use `np.random.default_rng`

_Originally posted by @ymzayek in https://github.com/nilearn/nilearn/pull/4054#discussion_r1358490487_

### PR DESCRIPTION
**Pull request by ymzayek**
- Closes #4055

Changes proposed in this pull request:

- Update Numpy RandomState to Numpy Generator using `numpy.random.default_rng`
  -  The Generator has a similar API to RandomState with some methods having a different name or different signature. [Numpy's API docs](https://numpy.org/doc/stable/reference/random/generator.html) have recommendations for what new methods replace ones in RandomState
- I did not yet update all instances of RandomState in the code but mostly stuck to our tests. The problem comes from some of our functions (e.g. in decomposition, decoding, massunivariate) passing RandomState to scikit learn functions which do not accept the Generator yet so it will take a bit more care to make sure we don't pass these where we call those functions. It could work to just pass a seed as int in some testing cases as scikit learn uses `sklearn.utils.check_random_state` internally to instantiate a RandomState object.

### CODE REVIEW

**Comment by ymzayek** [comment]
I updated the codebase where possible to use the new generator. In `decomposition` primarily this is not possible because for CanICA and DictLearning you can pass a RandomState object to their parameter random_state so this will break user code and will need proper deprecation.

**Comment by Remi-Gau** [comment]
> I updated the codebase where possible to use the new generator. In `decomposition` primarily this is not possible because for CanICA and DictLearning you can pass a RandomState object to their parameter random_state so this will break user code and will need proper deprecation.

I have not looked at it in detail but could we not just expand what types are passed to support also more "modern" rng options?

**Comment by Remi-Gau** [comment]
reread the top message. ignore what I just wrote.

**Comment by ymzayek** [comment]
Yes I should've been more clear. And it turns out it only happens in the public API that you can pass RandomState in the cases where it is eventually passed to a sklearn function that uses `check_random_state` internally. There are open issues on scikit learn github about this so I guess we can see how that develops but I have not read through them yet. https://github.com/scikit-learn/scikit-learn/issues/16988 and https://github.com/scikit-learn/scikit-learn/issues/20669

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Review by Remi-Gau — APPROVED** [review]
Except for one change I don't understand, the rest is good with me.

Thanks for taking of this one @ymzayek

File: nilearn/decoding/tests/test_decoder.py [hunk 1]
```diff
@@ -979,7 +980,7 @@ def test_decoder_multiclass_error_incorrect_cv(multiclass_data):
 
 def test_decoder_multiclass_warnings(multiclass_data, rng):
     X, y, _ = multiclass_data
-    groups = rng.binomial(2, 0.3, size=len(y))
```
[the remark this answers is not in the document]
Comment (ymzayek) [reply nilearn/decoding/tests/test_decoder.py:982]: @Remi-Gau in some test cases the random numbers generated differ enough between the 2 strategies leading to a failure and so they need a different seed. So in cases where I used the function with an explicit seed like this is because the fixture's seed of 42 fails the test

[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/decoding/tests/test_decoder.py:982]: ok but then we can remove rng from the list of arguments, no?
I may need more coffee but I don't see it being called anywhere else in the test.

[the remark this answers is not in the document]
Comment (ymzayek) [reply nilearn/decoding/tests/test_decoder.py:982]: Looks like I need more coffee haha
